File: leetcode/q297.java
package leetcode;

public class q297 {

// Definition for a binary tree node.
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        encode(root, sb);

        int len = sb.length();
        sb.deleteCharAt(len-1);
        System.out.println(sb.toString());
        
        return sb.toString();
    }

    private void encode(TreeNode node, StringBuilder sb) {
        if(node != null) {
            sb.append(node.val);
            sb.append(",");

            encode(node.left, sb);
            encode(node.right, sb);
        } else {
            sb.append("X");
            sb.append(",");
        }
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] data_arr = data.split(",");
        return decode(data_arr);
    }

    int index = 0;
    private TreeNode decode(String[] data_arr) {
        if(index == data_arr.length) return null;
        String curNode = data_arr[index++];
        if(curNode.equals("X")) return null;
        TreeNode root = new TreeNode(Integer.parseInt(curNode));
        root.left = decode(data_arr);
        root.right = decode(data_arr);

        return root;

    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// TreeNode ans = deser.deserialize(ser.serialize(root));
}

File: leetcode/q297.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:
    def encode(self, node, arr):
        if node is None:
            arr.append("X")
        else:
            arr.append(str(node.val))
            self.encode(node.left, arr)
            self.encode(node.right, arr)

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        arr = []
        self.encode(root, arr)
        print(arr)
        return ",".join(arr)           

    def decode(self, arr):
        if self.i < len(arr):
            cur = arr[self.i]
            self.i = self.i+1

            if cur == "X":
                return None
            
            node = TreeNode(int(cur))
            node.left = self.decode(arr)
            node.right = self.decode(arr)
            return node
        
        return None

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        arr = data.split(",")
        self.i = 0

        return self.decode(arr)

# Your Codec object will be instantiated and called as such:
ser = Codec()
deser = Codec()
code = "1,2,X,X,3,4,X,X,5,X,X"
ans = ser.serialize(deser.deserialize(code))
print(ans == code)